File: Shared/src/main/java/org/turbomedia/turboedit/shared/project/ProjectWriter.java
package org.turbomedia.turboedit.shared.project;

import org.msgpack.core.MessagePack;
import org.msgpack.core.MessagePacker;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

public class ProjectWriter {
    public static void Write(Project project) throws IOException {
        var file = new File(project.path());

        if (!file.exists()) {
            if (!file.createNewFile()) {
                throw new RuntimeException("Could not create Project file");
            }
        }

        var stream = new FileOutputStream(file);
        var packer = MessagePack.newDefaultPacker(stream);

        packer.packInt(project.fileVersion());
        packer.packString(project.name());

        // Files
        packer.packArrayHeader(project.files().size());
        for (var x = 0; x < project.files().size(); x++) {
            writeProjectFile(packer, project.files().get(x));
        }

        // Timelines
        packer.packArrayHeader(project.timelines().size());
        for (var x = 0; x < project.timelines().size(); x++) {
            writeTimelineData(packer, project.timelines().get(x));
        }

        packer.close();
    }

    private static void writeProjectFile(MessagePacker packer, ProjectFile data) throws IOException {
        packer.packString(data.name());
        packer.packString(data.path());
        packer.packString(data.type());

        // Preview Image
        packer.packArrayHeader(data.previewImage().length);
        for (var b = 0; b < data.previewImage().length; b++) {
            packer.packByte(data.previewImage()[b]);
        }

        packer.packBoolean(data.isVideo());

        if (data.isVideo()) {
            packer.packInt(data.videoHeight());
            packer.packInt(data.videoWidth());
            packer.packInt(data.videoFPS());

            // Audio Object
            packer.packArrayHeader(data.audio().size());
            for (var x = 0; x < data.audio().size(); x++) {
                writeAudioObject(packer, data.audio().get(x));
            }
        }
    }

    private static void writeAudioObject(MessagePacker packer, AudioObject data) throws IOException {
        packer.packInt(data.layer());
        packer.packString(data.layerName());

        // Waveform
        packer.packArrayHeader(data.waveform().length);
        for (var b = 0; b < data.waveform().length; b++) {
            packer.packByte(data.waveform()[b]);
        }
    }

    private static void writeTimelineData(MessagePacker packer, TimelineData data) throws IOException {
        packer.packString(data.name());
        packer.packLong(data.length());

        // Video Layers
        packer.packArrayHeader(data.videoLayers().size());
        for (var x = 0; x < data.videoLayers().size(); x++) {
            writeTimelineLayer(packer, data.videoLayers().get(x));
        }

        // Audio Layers
        packer.packArrayHeader(data.audioLayers().size());
        for (var x = 0; x < data.audioLayers().size(); x++) {
            writeTimelineLayer(packer, data.audioLayers().get(x));
        }

        // Clips
        packer.packArrayHeader(data.clips().size());
        for (var x = 0; x < data.clips().size(); x++) {
            writeTimelineClip(packer, data.clips().get(x));
        }
    }

    private static void writeTimelineLayer(MessagePacker packer, TimelineLayer data) throws IOException {
        packer.packString(data.name());
        packer.packBoolean(data.enabled());
    }

    private static void writeTimelineClip(MessagePacker packer, TimelineClip data) throws IOException {
        packer.packInt(data.file());
        packer.packInt(data.position());
        packer.packLong(data.startTime());
        packer.packLong(data.endTime());
    }
}

File: Shared/src/main/kotlin/turboedit/shared/project/ProjectWriter.kt
package turboedit.shared.project

import org.msgpack.core.MessagePack
import org.msgpack.core.MessagePacker
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

object ProjectWriter {
    @Throws(IOException::class)
    fun write(project: Project) {
        val file = File(project.path)

        if (!file.exists()) {
            if (!file.createNewFile()) {
                throw RuntimeException("Could not create Project file")
            }
        }

        val stream = FileOutputStream(file)
        val packer = MessagePack.newDefaultPacker(stream)

        packer.packInt(project.fileVersion)
        packer.packString(project.name)

        // Files
        packer.packArrayHeader(project.files.size)
        for (x in project.files.indices) {
            writeProjectFile(packer, project.files[x])
        }

        // Timelines
        packer.packArrayHeader(project.timelines.size)
        for (x in project.timelines.indices) {
            writeTimelineData(packer, project.timelines[x])
        }

        packer.close()
    }

    @Throws(IOException::class)
    private fun writeProjectFile(packer: MessagePacker, data: ProjectFile) {
        packer.packString(data.name)
        packer.packString(data.path)
        packer.packString(data.type)

        // Preview Image
        packer.packArrayHeader(data.previewImage?.size ?: 0)
        for (b in data.previewImage!!.indices) {
            packer.packByte(data.previewImage[b])
        }

        packer.packString(data.hash)
        packer.packLong(data.size)
        packer.packBoolean(data.isVideo)

        if (data.isVideo) {
            packer.packInt(data.videoWidth!!)
            packer.packInt(data.videoHeight!!)
            packer.packInt(data.videoFPS!!)
            packer.packString(data.videoCodec)

            // Audio Object
            packer.packArrayHeader(data.audio!!.size)
            for (x in data.audio.indices) {
                writeAudioObject(packer, data.audio[x])
            }
        }
    }

    @Throws(IOException::class)
    private fun writeAudioObject(packer: MessagePacker, data: AudioObject) {
        packer.packInt(data.layer)
        packer.packString(data.layerName)

        // Waveform
        packer.packArrayHeader(data.waveform.size)
        for (b in data.waveform.indices) {
            packer.packByte(data.waveform[b])
        }
    }

    @Throws(IOException::class)
    private fun writeTimelineData(packer: MessagePacker, data: TimelineData) {
        packer.packString(data.name)
        packer.packLong(data.length)

        // Video Layers
        packer.packArrayHeader(data.videoLayers.size)
        for (x in data.videoLayers.indices) {
            writeTimelineLayer(packer, data.videoLayers[x])
        }

        // Audio Layers
        packer.packArrayHeader(data.audioLayers.size)
        for (x in data.audioLayers.indices) {
            writeTimelineLayer(packer, data.audioLayers[x])
        }

        // Clips
        packer.packArrayHeader(data.clips.size)
        for (x in data.clips.indices) {
            writeTimelineClip(packer, data.clips[x])
        }
    }

    @Throws(IOException::class)
    private fun writeTimelineLayer(packer: MessagePacker, data: TimelineLayer) {
        packer.packString(data.name)
        packer.packBoolean(data.enabled)
    }

    @Throws(IOException::class)
    private fun writeTimelineClip(packer: MessagePacker, data: TimelineClip) {
        packer.packInt(data.file)
        packer.packLong(data.position)
        packer.packLong(data.startTime)
        packer.packLong(data.endTime)
    }
}